File: go/reference_contract_test.go
package yir

import (
	"encoding/json"
	"fmt"
	"testing"
)

func TestReferenceRoleContract(t *testing.T) {
	for _, tc := range []struct {
		images, videos, audios int
		valid                  bool
	}{
		{9, 3, 3, true}, {10, 0, 0, false}, {1, 4, 0, false}, {1, 0, 4, false}, {0, 0, 1, false},
	} {
		r := GenerationRequest{Model: "bytedance/seedance-2", Input: GenerationInput{Type: "reference", Prompt: "fixture"}}
		for role, count := range map[string]int{"reference_image": tc.images, "reference_video": tc.videos, "reference_audio": tc.audios} {
			for i := 0; i < count; i++ {
				r.Input.References = append(r.Input.References, Reference{Role: role, URL: fmt.Sprintf("https://example.com/%s/%d", role, i)})
			}
		}
		if err := ValidateGeneration("generate_video", r); (err == nil) != tc.valid {
			t.Fatalf("%+v: %v", tc, err)
		}
	}
}

func TestReferenceOutputDurationContract(t *testing.T) {
	for _, tc := range []struct {
		role     string
		duration int
		valid    bool
	}{{"reference_video", 10, true}, {"reference_video", 11, false}, {"reference_image", 11, true}} {
		r := GenerationRequest{Model: "alibaba/wan-2.7", Input: GenerationInput{Type: "reference", Prompt: "fixture", References: []Reference{{Role: tc.role, URL: "https://example.com/fixture"}}}, Parameters: map[string]any{"duration": tc.duration}}
		if err := ValidateGeneration("generate_video", r); (err == nil) != tc.valid {
			t.Fatalf("%+v: %v", tc, err)
		}
	}
}

func TestReferenceDuplicatesAndContractIsolation(t *testing.T) {
	reference := Reference{Role: "reference_image", URL: "https://example.com/fixture"}
	request := GenerationRequest{Model: "bytedance/seedance-2", Input: GenerationInput{Type: "reference", Prompt: "fixture", References: []Reference{reference, reference}}}
	if err := ValidateGeneration("generate_video", request); err == nil {
		t.Fatal("duplicate references accepted")
	}
	contract, ok := GetModelOperationContract(request.Model, "generate_video", "reference")
	if !ok {
		t.Fatal("missing contract")
	}
	constraint := contract.InputConstraints["reference"]
	constraint.ReferenceCountsByRole["reference_image"] = ReferenceCountRange{Maximum: 100}
	constraint.RequiredAnyReferenceRoles[0] = "invalid"
	fresh, _ := GetModelOperationContract(request.Model, "generate_video", "reference")
	if fresh.InputConstraints["reference"].ReferenceCountsByRole["reference_image"].Maximum != 9 || fresh.InputConstraints["reference"].RequiredAnyReferenceRoles[0] == "invalid" {
		t.Fatal("returned contract mutation escaped into the bundled catalog")
	}
}

func TestH3ReferenceRoleContract(t *testing.T) {
	image, _ := GetModelOperationContract("minimax/minimax-h3", "generate_video", "image")
	for _, rule := range image.Parameters {
		if rule.Name == "aspect_ratio" && (len(rule.Values) != 1 || rule.Values[0] != "adaptive") {
			t.Fatal("image aspect ratio expanded")
		}
	}
	for _, tc := range []struct {
		images, videos, audios int
		valid                  bool
	}{{9, 0, 0, true}, {0, 3, 0, true}, {0, 0, 3, true}, {9, 3, 3, true}, {10, 0, 0, false}, {1, 4, 0, false}, {1, 0, 4, false}, {0, 0, 0, false}} {
		r := GenerationRequest{Model: "minimax/minimax-h3", Input: GenerationInput{Type: "reference", Prompt: "fixture"}}
		for _, group := range []struct {
			role  string
			count int
		}{{"reference_audio", tc.audios}, {"reference_image", tc.images}, {"reference_video", tc.videos}} {
			for i := 0; i < group.count; i++ {
				r.Input.References = append(r.Input.References, Reference{Role: group.role, URL: fmt.Sprintf("https://example.com/%s/%d", group.role, i)})
			}
		}
		before, _ := json.Marshal(r)
		if err := ValidateGeneration("generate_video", r); (err == nil) != tc.valid {
			t.Fatalf("%+v: %v", tc, err)
		}
		after, _ := json.Marshal(r)
		if string(before) != string(after) {
			t.Fatal("request mutated")
		}
	}
}

File: typescript/tests/reference-contract.test.mjs
import assert from "node:assert/strict";
import test from "node:test";
import { validateGeneration, getModelOperationContract } from "../dist/index.js";

const ref = role => ({ role, url: "https://example.com/fixture" });
test("reference role counts and alternatives are exported and enforced", () => {
  const contract = getModelOperationContract("bytedance/seedance-2", "generate_video", "reference");
  assert.equal(contract.input_constraints.reference.reference_counts_by_role.reference_image.maximum, 9);
  for (const [images, videos, audios, valid] of [[9, 3, 3, true], [10, 0, 0, false], [1, 4, 0, false], [1, 0, 4, false], [0, 0, 1, false]]) {
    const references = [
      ...Array.from({ length: images }, (_, i) => ({ ...ref("reference_image"), url: `https://example.com/image/${i}` })),
      ...Array.from({ length: videos }, (_, i) => ({ ...ref("reference_video"), url: `https://example.com/video/${i}` })),
      ...Array.from({ length: audios }, (_, i) => ({ ...ref("reference_audio"), url: `https://example.com/audio/${i}` })),
    ];
    const check = () => validateGeneration("generate_video", { model: "bytedance/seedance-2", input: { type: "reference", prompt: "fixture", references }, parameters: {} });
    if (valid) assert.doesNotThrow(check); else assert.throws(check);
  }
});

test("reference role duration is output duration and applies only to the matching role", () => {
  for (const [role, duration, valid] of [["reference_video", 10, true], ["reference_video", 11, false], ["reference_image", 11, true]]) {
    const check = () => validateGeneration("generate_video", { model: "alibaba/wan-2.7", input: { type: "reference", prompt: "fixture", references: [ref(role)] }, parameters: { duration } });
    if (valid) assert.doesNotThrow(check); else assert.throws(check, error => error.code === "reference_duration_limit");
  }
});

test("duplicate references fail before transport", () => {
  assert.throws(() => validateGeneration("generate_video", { model: "bytedance/seedance-2", input: { type: "reference", prompt: "fixture", references: [ref("reference_image"), ref("reference_image")] }, parameters: {} }), error => error.code === "duplicate_reference");
});

test("H3 accepts official reference combinations without changing URLs or order", () => {
  const image = getModelOperationContract("minimax/minimax-h3", "generate_video", "image");
  assert.deepEqual(image.parameters.find(p => p.name === "aspect_ratio").values, ["adaptive"]);
  for (const [images, videos, audios, valid] of [[1, 0, 0, true], [9, 0, 0, true], [0, 3, 0, true], [0, 0, 3, true], [9, 3, 3, true], [10, 0, 0, false], [1, 4, 0, false], [1, 0, 4, false], [0, 0, 0, false]]) {
    const references = [["reference_audio", audios], ["reference_image", images], ["reference_video", videos]].flatMap(([role, count]) => Array.from({ length: count }, (_, i) => ({ role, url: `https://example.com/${role}/${i}` })));
    const before = structuredClone(references);
    const check = () => validateGeneration("generate_video", { model: "minimax/minimax-h3", input: { type: "reference", prompt: "fixture", references }, parameters: {} });
    if (valid) assert.doesNotThrow(check); else assert.throws(check);
    assert.deepEqual(references, before);
  }
});
